ruby/ql/lib/codeql/ruby/frameworks/ActionController.qll

```diff
@@ -357,3 +357,123 @@ private class ActionControllerProtectFromForgeryCall extends CsrfProtectionSetti
     if this.getWithValueText() = "exception" then result = true else result = false
   }
 }
+
+private module ParamsSummaries {
+  private import codeql.ruby.dataflow.FlowSummary
+
+  /**
+   * An instance of `ActionController::Parameters`, including those returned
+   * from method calls on other instances.
+   */
+  private class ParamsInstance extends DataFlow::Node {
+    ParamsInstance() {
+      this.asExpr().getExpr() instanceof ParamsCall
+      or
+      exists(DataFlow::CallNode call | call = this |
+        call.getReceiver() instanceof ParamsInstance and
+        call.getMethodName() = paramsMethodReturningParamsInstance()
+      )
+      or
+      exists(DataFlow::LocalSourceNode prev | prev.flowsTo(this))
+    }
+  }
```

Comment:
That is inevitable, I'm afraid. If `ParamsCall` depends on API graphs, shouldn't you already have an API node for it? In that case, using the API graph members will have to be good enough.

Reply:
Actually on second thought it seems really easy to add https://github.com/github/codeql/pull/10603. Would this work for you?

Reply:
OK, I've taken a look at that model and the problem arises when we're using API graphs to populate concepts that are based on AST nodes. Having to step outside API graphs should be considered tech debt, and not be a common case, and I believe Ruby's API graphs need to be developed a bit further for this to work. In particular we'll need to add `MethodDefinitionNode` (in JS this would correspond to a `MkDef` wrapping a `DataFlow::FunctionNode`) and something like the `MkClassInstance` node from JS. I'll look into that, but it will not arrive in time for this PR.